データベースアプライアンスのPatch実行時の型不一致

## 問題/症状

Patch funcを呼んだ際に以下のようなエラーとなる。

```console
panic: reflect.Set: value of type *sacloud.DatabaseSettingCommon is not assignable to type *sacloud.DatabaseSettingCommonUpdate
```

## 原因

PatchリクエストではRead -> 値のマッピング -> Updateを`sacloud.xxxOp`内で実装することで実現している。
(参考: https://github.com/sacloud/libsacloud/pull/316)

この時、Readで得られた値をmapconvでUpdate用のパラメータにマッピングするが、マッピング元/マッピング先で型が異なるためエラーとなっている。

型が異なるのは`WebUI`フィールドがCreate時と以外の時で型が異なるために違うパラメータ型としたため。

- Create時: bool(有効/無効)
- 以外: 文字列(WebUIのURL)

## 対応

今回は`WebUI`フィールド用に型を追加しRead/Updateで型を揃えるようにする。  

DSLを修正してRead/Updateで型が違う場合に対応しても良いが複雑さが増すため今回はDSLの修正は行わない。今後類似ケースで対応が必要になったら再度検討する。

--- a/internal/define/database.go
+++ b/internal/define/database.go
@@ -181,7 +181,7 @@ var (
 			fields.IconID(),
 
 			// settings
-			fields.DatabaseSettingsCommonUpdate(),
+			fields.DatabaseSettingsCommon(),
 			fields.DatabaseSettingsBackup(),
 			fields.DatabaseSettingsReplication(),
 			// settings hash

--- a/internal/define/fields.go
+++ b/internal/define/fields.go
@@ -1811,29 +1811,7 @@ func (f *fieldsDef) DatabaseSettingsCommon() *dsl.FieldDesc {
 			Name:      "DatabaseSettingCommon",
 			NakedType: meta.Static(naked.DatabaseSettingCommon{}),
 			Fields: []*dsl.FieldDesc{
-				fields.Def("WebUI", meta.TypeFlag),
-				fields.Def("ServicePort", meta.TypeInt),
-				fields.Def("SourceNetwork", meta.TypeStringSlice),
-				fields.Def("DefaultUser", meta.TypeString),
-				fields.Def("UserPassword", meta.TypeString),
-				fields.Def("ReplicaUser", meta.TypeString),
-				fields.Def("ReplicaPassword", meta.TypeString),
-			},
-		},
-		Tags: &dsl.FieldTags{
-			MapConv: "Settings.DBConf.Common,recursive",
-		},
-	}
-}
-
-func (f *fieldsDef) DatabaseSettingsCommonUpdate() *dsl.FieldDesc {
-	return &dsl.FieldDesc{
-		Name: "CommonSetting",
-		Type: &dsl.Model{
-			Name:      "DatabaseSettingCommonUpdate",
-			NakedType: meta.Static(naked.DatabaseSettingCommon{}),
-			Fields: []*dsl.FieldDesc{
-				fields.Def("WebUI", meta.TypeFlag), // TODO trueにした場合、文字列が返ってくるが未対応
+				fields.Def("WebUI", meta.Static(types.WebUI(""))),
 				fields.Def("ServicePort", meta.TypeInt),
 				fields.Def("SourceNetwork", meta.TypeStringSlice),
 				fields.Def("DefaultUser", meta.TypeString),

--- a/sacloud/test/database_op_test.go
+++ b/sacloud/test/database_op_test.go
@@ -150,7 +150,7 @@ var (
 	}
 	patchDatabaseParam = &sacloud.DatabasePatchRequest{
 		PatchEmptyToDescription: true,
-		CommonSetting: &sacloud.DatabaseSettingCommonUpdate{
+		CommonSetting: &sacloud.DatabaseSettingCommon{
 			ServicePort: 54321,
 		},
 	}
@@ -177,7 +177,7 @@ var (
 		Name:        testutil.ResourceName("db-upd"),
 		Tags:        []string{"tag1-upd", "tag2-upd"},
 		Description: "desc-upd",
-		CommonSetting: &sacloud.DatabaseSettingCommonUpdate{
+		CommonSetting: &sacloud.DatabaseSettingCommon{
 			ServicePort:     5432,
 			DefaultUser:     "exa.mple",
 			UserPassword:    "LibsacloudExamplePassword02",
@@ -219,7 +219,7 @@ var (
 				types.BackupSpanWeekdays.Monday,
 			},
 		},
-		CommonSetting: &sacloud.DatabaseSettingCommonUpdate{
+		CommonSetting: &sacloud.DatabaseSettingCommon{
 			ServicePort:     54321,
 			DefaultUser:     "exa.mple",
 			UserPassword:    "LibsacloudExamplePassword03",
@@ -256,7 +256,7 @@ var (
 	}
 	updateDatabaseToMinParam = &sacloud.DatabaseUpdateRequest{
 		Name: testutil.ResourceName("db-to-min"),
-		CommonSetting: &sacloud.DatabaseSettingCommonUpdate{
+		CommonSetting: &sacloud.DatabaseSettingCommon{
 			DefaultUser:     "exa.mple",
 			UserPassword:    "LibsacloudExamplePassword04",
 			ReplicaUser:     "replica",

--- a/sacloud/types/web_ui.go
+++ b/sacloud/types/web_ui.go
@@ -1,0 +1,61 @@
+// Copyright 2016-2019 The Libsacloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import "encoding/json"
+
+// WebUI データベースアプライアンスでのUI設定
+type WebUI string
+
+// String WebUIの文字列表現
+func (w *WebUI) String() string {
+	return string(*w)
+}
+
+// Bool WebUIの有効/無効
+func (w *WebUI) Bool() bool {
+	s := w.String()
+	if s == "" || s == "false" {
+		return false
+	}
+	return true
+}
+
+// MarshalJSON boolとstring両方に対応するための実装
+func (w WebUI) MarshalJSON() ([]byte, error) {
+	s := string(w)
+	switch s {
+	case "true", `1`, `"1"`:
+		return json.Marshal(true)
+	case "", `""`, `0`, `"0"`, "false":
+		return json.Marshal(false)
+	default:
+		return json.Marshal(s)
+	}
+}
+
+// UnmarshalJSON 文字列/boolが混在することへの対応
+func (w *WebUI) UnmarshalJSON(b []byte) error {
+	s := string(b)
+	switch s {
+	case "true", `1`, `"1"`:
+		*w = WebUI("true")
+	case "", `""`, `0`, `"0"`, "false":
+		*w = WebUI("false")
+	default:
+		*w = WebUI(s)
+	}
+	return nil
+}

--- a/sacloud/types/web_ui_test.go
+++ b/sacloud/types/web_ui_test.go
@@ -1,0 +1,91 @@
+// Copyright 2016-2019 The Libsacloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestWebUI_MarshalJSON(t *testing.T) {
+	cases := []struct {
+		in  WebUI
+		out []byte
+	}{
+		{
+			in:  WebUI(""),
+			out: []byte(`false`),
+		},
+		{
+			in:  WebUI("true"),
+			out: []byte(`true`),
+		},
+		{
+			in:  WebUI("false"),
+			out: []byte(`false`),
+		},
+		{
+			in:  WebUI("http://localhost:8080"),
+			out: []byte(`"http://localhost:8080"`),
+		},
+	}
+
+	for _, cc := range cases {
+		data, err := json.Marshal(cc.in)
+		if err != nil {
+			t.Fatal(err, cc)
+		}
+		require.Equal(t, cc.out, data, "target: %#v", cc)
+	}
+}
+
+func TestWebUI_UnmarshalJSON(t *testing.T) {
+	cases := []struct {
+		in      string
+		out     WebUI
+		enabled bool
+	}{
+		{
+			in:      `""`,
+			out:     WebUI(""),
+			enabled: false,
+		},
+		{
+			in:      `true`,
+			out:     WebUI("true"),
+			enabled: true,
+		},
+		{
+			in:      `false`,
+			out:     WebUI("false"),
+			enabled: false,
+		},
+		{
+			in:      `"http://localhost:8080"`,
+			out:     WebUI("http://localhost:8080"),
+			enabled: true,
+		},
+	}
+
+	for _, cc := range cases {
+		var v WebUI
+		if err := json.Unmarshal([]byte(cc.in), &v); err != nil {
+			t.Fatal(err, cc)
+		}
+		require.Equal(t, cc.enabled, v.Bool(), "target: %#v", cc)
+	}
+}

--- a/sacloud/zz_models.go
+++ b/sacloud/zz_models.go
@@ -4220,7 +4220,7 @@ func (o *Database) SetInterfaces(v []*InterfaceView) {
 
 // DatabaseSettingCommon represents API parameter/response structure
 type DatabaseSettingCommon struct {
-	WebUI           bool
+	WebUI           types.WebUI
 	ServicePort     int
 	SourceNetwork   []string
 	DefaultUser     string
@@ -4237,7 +4237,7 @@ func (o *DatabaseSettingCommon) Validate() error {
 // setDefaults implements sacloud.argumentDefaulter
 func (o *DatabaseSettingCommon) setDefaults() interface{} {
 	return &struct {
-		WebUI           bool
+		WebUI           types.WebUI
 		ServicePort     int
 		SourceNetwork   []string
 		DefaultUser     string
@@ -4256,12 +4256,12 @@ func (o *DatabaseSettingCommon) setDefaults() interface{} {
 }
 
 // GetWebUI returns value of WebUI
-func (o *DatabaseSettingCommon) GetWebUI() bool {
+func (o *DatabaseSettingCommon) GetWebUI() types.WebUI {
 	return o.WebUI
 }
 
 // SetWebUI sets value to WebUI
-func (o *DatabaseSettingCommon) SetWebUI(v bool) {
+func (o *DatabaseSettingCommon) SetWebUI(v types.WebUI) {
 	o.WebUI = v
 }
 
@@ -5074,11 +5074,11 @@ type DatabaseUpdateRequest struct {
 	Name               string `validate:"required"`
 	Description        string `validate:"min=0,max=512"`
 	Tags               types.Tags
-	IconID             types.ID                     `mapconv:"Icon.ID"`
-	CommonSetting      *DatabaseSettingCommonUpdate `mapconv:"Settings.DBConf.Common,recursive"`
-	BackupSetting      *DatabaseSettingBackup       `mapconv:"Settings.DBConf.Backup,recursive"`
-	ReplicationSetting *DatabaseReplicationSetting  `mapconv:"Settings.DBConf.Replication,recursive"`
-	SettingsHash       string                       `json:",omitempty" mapconv:",omitempty"`
+	IconID             types.ID                    `mapconv:"Icon.ID"`
+	CommonSetting      *DatabaseSettingCommon      `mapconv:"Settings.DBConf.Common,recursive"`
+	BackupSetting      *DatabaseSettingBackup      `mapconv:"Settings.DBConf.Backup,recursive"`
+	ReplicationSetting *DatabaseReplicationSetting `mapconv:"Settings.DBConf.Replication,recursive"`
+	SettingsHash       string                      `json:",omitempty" mapconv:",omitempty"`
 }
 
 // Validate validates by field tags
@@ -5092,11 +5092,11 @@ func (o *DatabaseUpdateRequest) setDefaults() interface{} {
 		Name               string `validate:"required"`
 		Description        string `validate:"min=0,max=512"`
 		Tags               types.Tags
-		IconID             types.ID                     `mapconv:"Icon.ID"`
-		CommonSetting      *DatabaseSettingCommonUpdate `mapconv:"Settings.DBConf.Common,recursive"`
-		BackupSetting      *DatabaseSettingBackup       `mapconv:"Settings.DBConf.Backup,recursive"`
-		ReplicationSetting *DatabaseReplicationSetting  `mapconv:"Settings.DBConf.Replication,recursive"`
-		SettingsHash       string                       `json:",omitempty" mapconv:",omitempty"`
+		IconID             types.ID                    `mapconv:"Icon.ID"`
+		CommonSetting      *DatabaseSettingCommon      `mapconv:"Settings.DBConf.Common,recursive"`
+		BackupSetting      *DatabaseSettingBackup      `mapconv:"Settings.DBConf.Backup,recursive"`
+		ReplicationSetting *DatabaseReplicationSetting `mapconv:"Settings.DBConf.Replication,recursive"`
+		SettingsHash       string                      `json:",omitempty" mapconv:",omitempty"`
 	}{
 		Name:               o.GetName(),
 		Description:        o.GetDescription(),
@@ -5170,12 +5170,12 @@ func (o *DatabaseUpdateRequest) SetIconID(v types.ID) {
 }
 
 // GetCommonSetting returns value of CommonSetting
-func (o *DatabaseUpdateRequest) GetCommonSetting() *DatabaseSettingCommonUpdate {
+func (o *DatabaseUpdateRequest) GetCommonSetting() *DatabaseSettingCommon {
 	return o.CommonSetting
 }
 
 // SetCommonSetting sets value to CommonSetting
-func (o *DatabaseUpdateRequest) SetCommonSetting(v *DatabaseSettingCommonUpdate) {
+func (o *DatabaseUpdateRequest) SetCommonSetting(v *DatabaseSettingCommon) {
 	o.CommonSetting = v
 }
 
@@ -5210,117 +5210,6 @@ func (o *DatabaseUpdateRequest) SetSettingsHash(v string) {
 }
 
 /*************************************************
-* DatabaseSettingCommonUpdate
-*************************************************/
-
-// DatabaseSettingCommonUpdate represents API parameter/response structure
-type DatabaseSettingCommonUpdate struct {
-	WebUI           bool
-	ServicePort     int
-	SourceNetwork   []string
-	DefaultUser     string
-	UserPassword    string
-	ReplicaUser     string
-	ReplicaPassword string
-}
-
-// Validate validates by field tags
-func (o *DatabaseSettingCommonUpdate) Validate() error {
-	return validator.New().Struct(o)
-}
-
-// setDefaults implements sacloud.argumentDefaulter
-func (o *DatabaseSettingCommonUpdate) setDefaults() interface{} {
-	return &struct {
-		WebUI           bool
-		ServicePort     int
-		SourceNetwork   []string
-		DefaultUser     string
-		UserPassword    string
-		ReplicaUser     string
-		ReplicaPassword string
-	}{
-		WebUI:           o.GetWebUI(),
-		ServicePort:     o.GetServicePort(),
-		SourceNetwork:   o.GetSourceNetwork(),
-		DefaultUser:     o.GetDefaultUser(),
-		UserPassword:    o.GetUserPassword(),
-		ReplicaUser:     o.GetReplicaUser(),
-		ReplicaPassword: o.GetReplicaPassword(),
-	}
-}
-
-// GetWebUI returns value of WebUI
-func (o *DatabaseSettingCommonUpdate) GetWebUI() bool {
-	return o.WebUI
-}
-
-// SetWebUI sets value to WebUI
-func (o *DatabaseSettingCommonUpdate) SetWebUI(v bool) {
-	o.WebUI = v
-}
-
-// GetServicePort returns value of ServicePort
-func (o *DatabaseSettingCommonUpdate) GetServicePort() int {
-	return o.ServicePort
-}
-
-// SetServicePort sets value to ServicePort
-func (o *DatabaseSettingCommonUpdate) SetServicePort(v int) {
-	o.ServicePort = v
-}
-
-// GetSourceNetwork returns value of SourceNetwork
-func (o *DatabaseSettingCommonUpdate) GetSourceNetwork() []string {
-	return o.SourceNetwork
-}
-
-// SetSourceNetwork sets value to SourceNetwork
-func (o *DatabaseSettingCommonUpdate) SetSourceNetwork(v []string) {
-	o.SourceNetwork = v
-}
-
-// GetDefaultUser returns value of DefaultUser
-func (o *DatabaseSettingCommonUpdate) GetDefaultUser() string {
-	return o.DefaultUser
-}
-
-// SetDefaultUser sets value to DefaultUser
-func (o *DatabaseSettingCommonUpdate) SetDefaultUser(v string) {
-	o.DefaultUser = v
-}
-
-// GetUserPassword returns value of UserPassword
-func (o *DatabaseSettingCommonUpdate) GetUserPassword() string {
-	return o.UserPassword
-}
-
-// SetUserPassword sets value to UserPassword
-func (o *DatabaseSettingCommonUpdate) SetUserPassword(v string) {
-	o.UserPassword = v
-}
-
-// GetReplicaUser returns value of ReplicaUser
-func (o *DatabaseSettingCommonUpdate) GetReplicaUser() string {
-	return o.ReplicaUser
-}
-
-// SetReplicaUser sets value to ReplicaUser
-func (o *DatabaseSettingCommonUpdate) SetReplicaUser(v string) {
-	o.ReplicaUser = v
-}
-
-// GetReplicaPassword returns value of ReplicaPassword
-func (o *DatabaseSettingCommonUpdate) GetReplicaPassword() string {
-	return o.ReplicaPassword
-}
-
-// SetReplicaPassword sets value to ReplicaPassword
-func (o *DatabaseSettingCommonUpdate) SetReplicaPassword(v string) {
-	o.ReplicaPassword = v
-}
-
-/*************************************************
 * DatabasePatchRequest
 *************************************************/
 
@@ -5333,7 +5222,7 @@ type DatabasePatchRequest struct {
 	PatchEmptyToTags               bool
 	IconID                         types.ID `mapconv:"Icon.ID"`
 	PatchEmptyToIconID             bool
-	CommonSetting                  *DatabaseSettingCommonUpdate `mapconv:"Settings.DBConf.Common,recursive"`
+	CommonSetting                  *DatabaseSettingCommon `mapconv:"Settings.DBConf.Common,recursive"`
 	PatchEmptyToCommonSetting      bool
 	BackupSetting                  *DatabaseSettingBackup `mapconv:"Settings.DBConf.Backup,recursive"`
 	PatchEmptyToBackupSetting      bool
@@ -5357,7 +5246,7 @@ func (o *DatabasePatchRequest) setDefaults() interface{} {
 		PatchEmptyToTags               bool
 		IconID                         types.ID `mapconv:"Icon.ID"`
 		PatchEmptyToIconID             bool
-		CommonSetting                  *DatabaseSettingCommonUpdate `mapconv:"Settings.DBConf.Common,recursive"`
+		CommonSetting                  *DatabaseSettingCommon `mapconv:"Settings.DBConf.Common,recursive"`
 		PatchEmptyToCommonSetting      bool
 		BackupSetting                  *DatabaseSettingBackup `mapconv:"Settings.DBConf.Backup,recursive"`
 		PatchEmptyToBackupSetting      bool
@@ -5473,12 +5362,12 @@ func (o *DatabasePatchRequest) SetPatchEmptyToIconID(v bool) {
 }
 
 // GetCommonSetting returns value of CommonSetting
-func (o *DatabasePatchRequest) GetCommonSetting() *DatabaseSettingCommonUpdate {
+func (o *DatabasePatchRequest) GetCommonSetting() *DatabaseSettingCommon {
 	return o.CommonSetting
 }
 
 // SetCommonSetting sets value to CommonSetting
-func (o *DatabasePatchRequest) SetCommonSetting(v *DatabaseSettingCommonUpdate) {
+func (o *DatabasePatchRequest) SetCommonSetting(v *DatabaseSettingCommon) {
 	o.CommonSetting = v
 }
 

